accepted/2021/Globalization/GlobalizationSize.md

```diff
@@ -0,0 +1,81 @@
+
+```
+ICU4X will provide an ECMA-402-compatible API surface in the target client-side platforms, including the web platform, iOS, Android, WearOS, WatchOS, Flutter, and Fuchsia, supported in programming languages including Rust, JavaScript, Objective-C, Java, Dart, and C++.
+
+The design goals of ICU4X are:
+- Small and modular code
+- Pluggable locale data
+- Availability and ease of use in multiple programming languages
+- Written by i18n experts to encourage best practices
+```
+
+That is exactly what we need (or what we are trying to achieve).
+Eric Erhardt already did some investigation and arranged a meeting with the project committee to learn more about the project and the status of the project. It was very helpful getting in touch and learning more about this project. It is a promising project managed by experts and using the CLDR data which means it still sticking with the standards.
+
+The catch here is ICU4X still in the early stages and still not providing all functionality needed by .NET for globalization support. The project committee is willing to get and prioritize requests from different parties. We already communicated in the meeting what functionality .NET currently using from ICU4C and want to see that supported in ICU4X. For example, collation support is one of the topics we brought to the committee's attention.
+
+### Plan Suggestion
+
+ICU4X is the promising path we should pursue to address all concerned scenarios. We can wait a little bit more to get the needed missing functionality implemented in the project and integrate it to .NET as we did with ICU4C.
+
+- We need to look more at what is currently supported functionality and what is missing so we can have the full list of features we need to get.
+- We need to be in touch with the ICU4X committee communicating our requests and understanding when such features can be available.
+- Need to look if we can associate some resources to help with that project especially with the missing features we need in the .NET.
+- Need to look at the scope of work to integrate ICU4X to .NET runtime.
+- Need to look if we can have a process creating different NuGet packages from the ICU4X repo with data and code customization.
+
+.NET 7.0 would be the best to start invest in that and try to start consuming at least the available parts of ICU4X.
+
+### Alternative Plans
+
+Other ideas would be similar to what the ICU4X is trying to do but maybe with scoped level. Here are the options we can try if we didn't go with ICU4X:
+
+- Invest more in ICU4C trimming. That will need spending more resources doing code analysis and figuring out how we may trim the ICU code to the level that can satisfy the size requirements.
+- Write a code wrapper around the CLDR or ICU data. So we'll not use ICU code to access the data. This can work for locale data but I don't think that will be a good option for other functionality like collation as the code is more complicated and not easy to re-implement.
```

Comment:
ICU is part of the Mac ecosystem. However we cannot directly use it, it's considered private (and using it would result in app rejections).

However it means that for macOS (including Catalyst), iOS, tvOS (and even watchOS) Apple provide their own API built on top of ICU (code and or data).

It might be possible to replace the globalization code to use those API (for the mentioned OS) instead of providing both ICU code and data. 

Since most of the code (and data) is part of the OS then the resulting app size should be reduced and closer to the "legacy" numbers. E.g. https://github.com/xamarin/xamarin-macios/issues/10249#issuecomment-854904528

Reply:
I believe the problem with this tends to be APIs not lining up or returning different data from what we've established as the standard on other platforms?

Reply:
If the Mac APIs at the end depend on the ICU, then I expect the behavior will be close as if we use ICU directly. 

@CoffeeFlux do you have more information about the differences?

Reply:
I had a very rough look at how the Cocoa API surface maps to the System.Globalization.Native API surface. It seems to be quite promising even if not a full match. The missing parts could probably be filled in with custom data.

GlobalizationNative_NormalizeString
- NSString decomposedStringWithCanonicalMapping (form D)
- NSString decomposedStringWithCompatibilityMapping (form KD)
- NSString precomposedStringWithCanonicalMapping (form C)
- NSString precomposedStringWithCompatibilityMapping (form KC)

GlobalizationNative_IsNormalized
- ?? (could be mapped to GlobalizationNative_NormalizeString but inefficient)

GlobalizationNative_WindowsIdToIanaId [Not supported on app-local iOS/WASM ICU]
GlobalizationNative_IanaIdToWindowsId [Not supported on app-local iOS/WASM ICU]

GlobalizationNative_GetTimeZoneDisplayName [Not supported on app-local iOS/WASM ICU]
- NSTimeZone

GlobalizationNative_GetLocaleInfoString
- NSLocale
  - LocaleString_LocalizedDisplayName => localizedStringForLocaleIdentifier
  - LocaleString_EnglishDisplayName => localeIdentifier / localizedStringForLocaleIdentifier?
  - LocaleString_NativeDisplayName => localeIdentifier / localizedStringForLocaleIdentifier?
  - LocaleString_LocalizedLanguageName => localizedStringForLanguageCode
  - LocaleString_EnglishLanguageName => localizedStringForLanguageCode
  - LocaleString_NativeLanguageName => localizedStringForLanguageCode
  - LocaleString_EnglishCountryName => localizedStringForCountryCode
  - LocaleString_NativeCountryName => localizedStringForCountryCode
  - LocaleString_DecimalSeparator => decimalSeparator
  - LocaleString_ThousandSeparator => groupingSeparator
  - LocaleString_Digits => ?
  - LocaleString_MonetarySymbol => currencySymbol
  - LocaleString_CurrencyEnglishName => localizedStringForCurrencyCode
  - LocaleString_CurrencyNativeName => localizedStringForCurrencyCode
  - LocaleString_Iso4217MonetarySymbol => currencySymbol (?)
  - LocaleString_MonetaryDecimalSeparator => ?
  - LocaleString_MonetaryThousandSeparator => ?
  - LocaleString_AMDesignator => ?
  - LocaleString_PMDesignator => ?
  - LocaleString_PositiveSign => ?
  - LocaleString_NegativeSign => ?
  - LocaleString_Iso639LanguageTwoLetterName => languageCode
  - LocaleString_Iso639LanguageThreeLetterName => ?
  - LocaleString_Iso3166CountryName => countryCode (?)
  - LocaleString_Iso3166CountryName2 => ?
  - LocaleString_NaNSymbol => ?
  - LocaleString_PositiveInfinitySymbol => ?
  - LocaleString_ParentName => ?
  - LocaleString_PercentSymbol => ? 
  - LocaleString_PerMilleSymbol => ?

GlobalizationNative_GetLocaleTimeFormat
- NSDateFormatter?

GlobalizationNative_GetLocaleInfoInt
- NSLocale

GlobalizationNative_GetLocaleInfoGroupingSizes
- ??

GlobalizationNative_GetLocales
GlobalizationNative_GetLocaleName
GlobalizationNative_GetDefaultLocaleName
GlobalizationNative_IsPredefinedLocale
- NSLocale

GlobalizationNative_ToAscii
GlobalizationNative_ToUnicode
- No equivalent API for IDN/Punycode?

GlobalizationNative_GetSortHandle
- return reference to NSLocale

GlobalizationNative_CloseSortHandle
- release reference to NSLocale

GlobalizationNative_GetSortVersion
- ??

GlobalizationNative_CompareString
- [NSString compare:options:range:locale:](https://developer.apple.com/documentation/foundation/nsstring/1414561-compare?language=objc)

GlobalizationNative_IndexOf
- [NSString rangeOfString:options:range:locale:](https://developer.apple.com/documentation/foundation/nsstring/1417348-rangeofstring?language=objc)

GlobalizationNative_LastIndexOf
- same as GlobalizationNative_IndexOf w/ NSBackwardsSearch

GlobalizationNative_StartsWith
- can be implemented trough GlobalizationNative_CompareString?

GlobalizationNative_EndsWith
- ?

GlobalizationNative_GetSortKey
- ?

GlobalizationNative_ChangeCase
- NSString uppercaseStringWithLocale
- NSString lowercaseStringWithLocale

GlobalizationNative_ChangeCaseInvariant
- NSString uppercaseString
- NSString lowercaseString

GlobalizationNative_ChangeCaseTurkish
- ?

GlobalizationNative_InitOrdinalCasingPage
- ?

GlobalizationNative_GetCalendars
GlobalizationNative_GetCalendarInfo
GlobalizationNative_EnumCalendarInfo
GlobalizationNative_GetLatestJapaneseEra
GlobalizationNative_GetJapaneseEraStartDate
- NSCalendar (TBD: details)

Reply:
That is a good input to have. This path is worth to explore more.

Reply:
Most importantly it seems to cover the collations which are hard to implement. The things I could not find at first glance seem to be simple data retrieval.

Reply:
@filipnavara I'll add some section to the doc for the option trying to use the Cocoa APIs for Apple OS's.

Reply:
> It might be possible to replace the globalization code to use those API (for the mentioned OS) instead of providing both ICU code and data.

Based on Mozilla experience, one consideration I must share is that doing that will lead to your software having diverging capabilities between platforms where you use ICU4C that you vendor in, vs platforms where you rely on the OS one.

ICU API does change, and in particular active large scale software often goes through a cycle where a need for an ICU4C is identitied, requested, supplied and then if the ICU4C can be vendored in, the next release contains new code. If you rely on OS ICU, the path may be extended to many years of wait which may be a late realization of a drag factor if you were to be affected.

Reply:
@zbraniecki I agree with your consideration. What we are trying to do here we weigh the options and see pros and cons of every option. But in general, if consistency across platforms and OS's is available option with reasonable cost for the scenarios we are targeting (e.g. mobile and WebAssembly scenarios) then that should be the option we go with.